iree/compiler/Dialect/Flow/Transforms/ConvertLinalgMatmulToMmt4D.cpp

```diff
@@ -18,50 +18,71 @@ namespace Flow {
 
 namespace {
 
-// Expands a 2d tensor operand to 4d given its target shape.
+// Expands a 2D tensor input to a 4D tensor representing the same underlying
+// data but now in a tiled layout, given a static 2D tile shape.
 // Does not transpose.
 // Example: (M, N) --> (M1, M0, N1, N0)
 static Value expandTo4D(mlir::Location loc, PatternRewriter &rewriter,
-                        Value operand, ArrayRef<int64_t> targetShape) {
-  auto operandType = operand.getType().cast<RankedTensorType>();
-  auto targetType =
-      RankedTensorType::get(targetShape, operandType.getElementType());
+                        Value input, ArrayRef<int64_t> tileShape) {
+  RankedTensorType inputType = input.getType().cast<RankedTensorType>();
+  ArrayRef<int64_t> inputStaticShape = inputType.getShape();
+  SmallVector<int64_t, 4> targetShape;
+  // Generate a 4D static shape of the form (M1, M0, N1, N0),
+  // where M0, N0 are always static and M1, N1 are static if and only if M, N
+  // are.
+  for (int i : {0, 1}) {
+    if (inputStaticShape[i] == ShapedType::kDynamicSize) {
+      targetShape.push_back(ShapedType::kDynamicSize);
+    } else {
+      targetShape.push_back(inputStaticShape[i] / tileShape[i]);
+    }
+    targetShape.push_back(tileShape[i]);
+  }
+  RankedTensorType targetType =
+      RankedTensorType::get(targetShape, inputType.getElementType());
   SmallVector<ReassociationIndices> expandIndices = {{0, 1}, {2, 3}};
   Value reshapedOperand = rewriter.create<linalg::TensorExpandShapeOp>(
-      loc, targetType, operand, expandIndices);
+      loc, targetType, input, expandIndices);
   return reshapedOperand;
 }
 
-// Creates a linalg.generic that transposes operand using permutation indices.
+// Creates a linalg.generic that transposes input using permutation indices.
 // Example: (M1, M0, N1, N0) -> (M1, N1, M0, N0) if indices = {0, 2, 1, 3}.
 static Value transpose(mlir::Location loc, PatternRewriter &rewriter,
-                       Value operand, ArrayRef<int64_t> indices) {
-  RankedTensorType operandTensorType =
-      operand.getType().cast<RankedTensorType>();
+                       Value input, ArrayRef<int64_t> indices) {
+  RankedTensorType inputType = input.getType().cast<RankedTensorType>();
   auto nloops = indices.size();
-  auto inputShape = operandTensorType.getShape();
 
   SmallVector<AffineExpr, 4> exprs = llvm::to_vector<4>(
```

Comment:
You can probably just use `AffineMap::getPermutationMap` [here](https://github.com/llvm/llvm-project/blob/5a4bd07ea4643fb9d1d11f690ca6c0e12f4cf6b5/mlir/include/mlir/IR/AffineMap.h#L81)

Reply:
ah thanks... this is a preexisting issue in this code. I will add a TODO. Thanks!